### PR TITLE
SOLR-13489: Stop the leader from trying to rejoin the election on ses…

### PR DESCRIPTION
**Pull request by markrmiller**
…sion expiration and harden our zk reconnect code path.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 1]
```diff
@@ -368,8 +368,15 @@ public void command() {
...
-
-              registerAllCoresAsDown(registerOnReconnect, false);
+              
+              try {
+                registerAllCoresAsDown(registerOnReconnect, false);
+              } catch (SessionExpiredException e) {
+                // zk has to reconnect
+                throw e;
+              } catch (Exception e) {
+                log.warn("Exception while trying to register all cores as DOWN", e);
```
Comment (anshumg) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: don't you want to re-throw the exception here?

Comment (markrmiller) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: I'll add a better comment to the code.

Comment (anshumg) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: I should've guessed :) I just thought that you missed handling more cases.

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 2]
```diff
@@ -1903,21 +1922,20 @@ private ZkCoreNodeProps waitForLeaderToSeeDownState(
             " is already less than leader, so not waiting for leader to see down state.");
       } else {
 
-        log.info("Replica need to wait for leader to see down state.");
+        log.info("Replica should to wait for leader to see down state.");
```
Comment (anshumg) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: Replica `is` waiting ?

Comment (anshumg) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: Also, while you are at it, can you add more information e.g. specifying the replica ?

File: solr/solrj/src/java/org/apache/solr/common/cloud/ConnectionManager.java [hunk 3]
```diff
@@ -155,27 +155,28 @@ public void process(WatchedEvent event) {
                 public void update(SolrZooKeeper keeper) {
                   try {
                     waitForConnected(Long.MAX_VALUE);
+                    
+                    log.info("Connection with ZooKeeper reestablished.");
```
Comment (anshumg) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ConnectionManager.java]: we already log this as debug in `waitForConnected`. Let's just switch that out to an info ?

**Pull request closed without merging**